src/rgw/rgw_log.cc

```diff
@@ -317,32 +317,70 @@ void rgw_format_ops_log_entry(struct rgw_log_entry& entry, Formatter *formatter)
   formatter->close_section();
 }
 
-void OpsLogSocket::formatter_to_bl(bufferlist& bl)
+OpsLogManifold::~OpsLogManifold()
 {
-  stringstream ss;
-  formatter->flush(ss);
-  const string& s = ss.str();
+    for (const auto &sink : sinks) {
+        delete sink;
+    }
+}
 
-  bl.append(s);
+void OpsLogManifold::add_sink(OpsLogSink *sink)
+{
+    sinks.push_back(sink);
 }
 
-void OpsLogSocket::init_connection(bufferlist& bl)
+int OpsLogManifold::log(struct req_state* s, struct rgw_log_entry& entry)
 {
-  bl.append("[");
+  int ret = 0;
+  for (const auto &sink : sinks) {
+    if (sink->log(s, entry) < 0) {
+      ret = -1;
+    }
+  }
+  return ret;
 }
 
-OpsLogSocket::OpsLogSocket(CephContext *cct, uint64_t _backlog) : OutputDataSocket(cct, _backlog)
+OpsLogFile::OpsLogFile(std::string& path) :
+  file(path, std::ofstream::app)
 {
-  formatter = new JSONFormatter;
-  delim.append(",\n");
 }
 
-OpsLogSocket::~OpsLogSocket()
+OpsLogFile::~OpsLogFile()
 {
+  file.close();
+}
+
+int OpsLogFile::log_json(struct req_state* s, bufferlist& bl)
+{
+  lock.lock();
+  bl.write_stream(file);
```

Comment:
@mattbenjamin Yeah that was a question that I intended to ask. For our use case, we don't want to lose these ops logs in the case of a process crash, so we prefer that they be written out synchronously. I can imagine that this might be less important for other users and that it might be desirable to have a knob to turn in order to control whether (and how much) buffering is used. Since you're confirming that this is important, I'll go ahead and work on that piece and update the PR.

Reply:
these synchronous writes are unfortunately in conflict with the async refactoring project, where we're trying to eliminate all blocking waits from frontend threads. beast frontend requests run as coroutines, so when we can suspend that coroutine instead of blocking, that allows the thread to process other requests in the meantime

i'd eventually like to see this use something like libaio to make the file writes asynchronous